python/ray/serve/controller.py

```diff
@@ -605,7 +549,6 @@ def shutdown(self):
                 extra={"log_to_stderr": False},
             )
             _controller_actor = ray.get_runtime_context().current_actor
-            self._shutdown.set()
```

Comment:
@edoakes @GeneDer This may be a behavior change, so please check this.

We set this event before `ray.killing` the controller actor. However, as far as I can tell, the only call that `awaits` this event is `graceful_shutdown`. The expected behavior for `graceful_shutdown` is that the caller waits until the controller dies, and the caller receives a `RayActorError`. There seems to be no reason to set this event since `graceful_shutdown` shouldn't return before the controller dies.

I had to remove this because there were a few tests in `test_standalone_3` that were failing (e.g. `test_controller_shutdown_gracefully`). They check that calling `graceful_shutdown` raises a `RayActorError`. These were passing because snapshotting the cluster state gave enough time for `ray.kill` to finish executing before the controller exited the control loop, which raised the `RayActorError` before `graceful_shutdown` could return. However, without snapshotting, the `graceful_shutdown` call was returning before the `RayActorError` got raised.

As further evidence that the tests failed due to a timing issue, I got the test to pass by replacing `self._put_serve_snapshot()` with `time.sleep(0.2)`.

Reply:
Hmm, I'm not sure if I understand the problem 😅 `graceful_shutdown` is called in the client to signal to the controller to shutdown itself with or without the client needing to wait for the process (since it's just setting a flag on the controller and controller will deal with the shutdown sequence in the event loop). `RayActorError` is a byproduct of the client waiting for the `graceful_shutdown` to return, but the actor is already dead right? Maybe instead of dropping this event, we can just drop the `RayActorError` error catching given it now returns faster than ray to kill the controller?

Reply:
`graceful_shutdown` takes in one argument: [`wait`](https://github.com/ray-project/ray/blob/f22259bbd676a9e433c5ef0887c2a5a32887cfef/python/ray/serve/controller.py#L1019). When `wait` is `True`, the client waits until the controller shuts down, which should raise a `RayActorError` (at least based on how [`test_controller_shutdown_gracefully`](https://github.com/ray-project/ray/blob/f22259bbd676a9e433c5ef0887c2a5a32887cfef/python/ray/serve/tests/test_standalone_3.py#L620-L621) is implemented).

We should clarify the expected behavior here. If `wait` is `True`, should the client wait until the Serve controller is dead? I think the best way to ensure that is to wait until the controller is actually dead and raises a `RayActorError`, instead of returning when the event is set.

Reply:
Oh I got your point! Yep, agreed that way we are sure the client can wait until controller is dead :)